brave/src/main/java/brave/internal/recorder/MutableEndpoint.java

```diff
@@ -0,0 +1,176 @@
+    this.serviceName = serviceName == null || serviceName.isEmpty() ? null : serviceName;
+  }
+
+  /**
+   * The text representation of the primary IPv4 or IPv6 address associated with this connection.
+   * Ex. 192.168.99.100 null if unknown.
+   */
+  @Nullable public String ip() {
+    return ip;
+  }
+
+  /** @see #ip */
+  public void ip(@Nullable String ip) {
+    if (ip == null || ip.isEmpty()) {
+      this.ip = null;
+      return;
+    }
+    IpFamily format = detectFamily(ip);
+    if (format == IpFamily.IPv4 || format == IpFamily.IPv6) {
+      this.ip = ip;
+    } else if (format == IpFamily.IPv4Embedded) {
+      this.ip = ip.substring(ip.lastIndexOf(':') + 1);
+    } else {
+      this.ip = null;
+    }
+  }
+
+  /**
+   * Port of the IP's socket or 0, if not known.
+   *
+   * @see java.net.InetSocketAddress#getPort()
+   */
+  public int port() {
+    return port;
+  }
+
+  /** @see MutableEndpoint#port() */
+  public void port(int port) {
+    if (port > 0xffff) throw new IllegalArgumentException("invalid port " + port);
+    if (port < 0) port = 0;
+    this.port = port;
+  }
+
+  @Override public String toString() {
+    return "MutableEndpoint{serviceName=" + serviceName + ", ip=" + ip + ", port=" + port + "}";
+  }
+
+  @Override public boolean equals(Object o) {
+    if (o == this) return true;
+    if (!(o instanceof MutableEndpoint)) return false;
+    MutableEndpoint that = (MutableEndpoint) o;
+    return ((serviceName == null)
+        ? (that.serviceName == null) : serviceName.equals(that.serviceName))
+        && ((ip == null) ? (that.ip == null) : ip.equals(that.ip))
+        && port == that.port;
+  }
+
+  @Override public int hashCode() {
+    int h = 1;
+    h *= 1000003;
+    h ^= (serviceName == null) ? 0 : serviceName.hashCode();
+    h *= 1000003;
+    h ^= (ip == null) ? 0 : ip.hashCode();
+    h *= 1000003;
+    h ^= port;
+    return h;
+  }
+
+  // All the below code is from zipkin2.Endpoint, copy/pasted here to prevent a depedency.
+  enum IpFamily {
+    Unknown,
+    IPv4,
+    IPv4Embedded,
+    IPv6
+  }
+
+  /**
+   * Adapted from code in {@code com.google.common.net.InetAddresses.ipStringToBytes}. This version
+   * separates detection from parsing and checks more carefully about embedded addresses.
+   */
+  static IpFamily detectFamily(String ipString) {
+    boolean hasColon = false;
+    boolean hasDot = false;
+    for (int i = 0, length = ipString.length(); i < length; i++) {
+      char c = ipString.charAt(i);
+      if (c == '.') {
+        hasDot = true;
+      } else if (c == ':') {
+        if (hasDot) return IpFamily.Unknown; // Colons must not appear after dots.
+        hasColon = true;
+      } else if (notHex(c)) {
+        return IpFamily.Unknown; // Everything else must be a decimal or hex digit.
+      }
+    }
+
+    // Now decide which address family to parse.
+    if (hasColon) {
+      if (hasDot) {
+        int lastColon = ipString.lastIndexOf(':');
```

Comment:
`lastColonPos` or `lastColonIndex` is maybe a better name. WDYT?

Reply:
copy/pasted code, but agreed it is a better name. I'll switch in both places.